Feat: achivements scene user feedback  (EOSU-653)

JessTello opened the pull request:
- Display updated achievement image right after unlocking for better user feedback
- Disable unlock button when achievement is already unlocked

JessTello committed:
Feat: achivements scene user feedback  (EOSU-653)

- Display updated achievement image right after unlocking for better user feedback
- Disable unlock button when achievement is already unlocked

matt-clarke commented on Assets/Scripts/StandardSamples/UI/Achievements/UIAchievementsMenu.cs:
```diff
@@ -87,20 +87,48 @@ public async void IncrementLoginStat()
+                    Progress = 1.0,
+                    UnlockTime = DateTime.UtcNow,
+                    StatInfo = null
+                };
```
The menu creating the PlayerAchievement here is a bit backwards, it shouldn't have to own that. I would update `UnlockAchievementAsync()` so the Task result is the `PlayerAchievement`

matt-clarke commented on Assets/Scripts/StandardSamples/UI/Achievements/UIAchievementsMenu.cs:
```diff
@@ -87,20 +87,48 @@ public async void IncrementLoginStat()
+                Texture2D lockedIcon = await AchievementsService.Instance.GetAchievementLockedIconTexture(definition.AchievementId);
+
+                achievementUnlockedIcon.texture = unlockedIcon;
+                achievementLockedIcon.texture = lockedIcon;
```
Is it necessary to retrieve the textures and assign them again? Wouldn't they already be set?

billfr-ApexSystems replied:
@JessTello Validate the need for texture assignment and determine if fixing original assignment should be targeted fix as well.

JessTello replied:
It is not being assigned twice to the same element; the first assignment corresponds to the thumbnail icon, and the second one to the main icon.
<img width="546" height="193" alt="image" src="https://github.com/user-attachments/assets/fa05bc3e-f484-46a2-b79b-95ea7b97b55d" />

JessTello replied:
Without this assignment, the main icon does not update and remains in a locked state, which prevents the expected behavior.

<img width="575" height="298" alt="image" src="https://github.com/user-attachments/assets/9fe3c8ea-6a2a-423a-b4f5-6f0a3165d043" />

matt-clarke replied:
With that in mind, one optimisation would be to retrieve the textures from the other icon and assign it to this icon. That eliminates the need to download the textures again.

If there's anything which makes that non-trivial then don't worry about it

JessTello replied:
Implemented this comment in the new commit

JessTello committed:
Refactor: Separete UI logic from Task

- Update UnlockAchievementAsync to return PlayerAchievement
- Limit UIAchievementsMenu to UI responsibilities

matt-clarke commented on Assets/Scripts/StandardSamples/Services/AchievementsService.cs:
```diff
@@ -493,23 +493,29 @@ public static uint GetAchievementsCount()
         /// The id of the achievement to unlock for the current player.
         /// </param>
-        public Task UnlockAchievementAsync(string achievementId)
+        public Task<PlayerAchievement> UnlockAchievementAsync(AchievementUnlockInfo info)
```
This class has all the data about the achievements, so it should only need the id passed in. Then you can loop over the `_achievements` list to find the data. If the data isn't found then you can set an exception on the Task.

JessTello replied:
I’ve already implemented this comment in the new commit. Thank you for your feedback!

JessTello committed:
Refactor: achievement unlock flow to support detailed info and async icon handling

- Updated UnlockAchievementAsync to accept AchievementUnlockInfo
- Improved logic for achievement definitions and player achievements list
- Implemented async retrieval of achievement icons
- Adjusted UI button interactable state based on lock status

matt-clarke commented on Assets/Scripts/StandardSamples/Services/AchievementsService.cs:
```diff
@@ -493,23 +493,46 @@ public static uint GetAchievementsCount()
+            public Utf8String AchievementId;
+            public string DisplayName;
+            public string Description;
+        }
```
This type is now unused and can be removed
```suggestion
```

JessTello replied:
Fixed in the latest commit.

matt-clarke commented on Assets/Scripts/StandardSamples/Services/AchievementsService.cs:
```diff
@@ -493,23 +493,46 @@ public static uint GetAchievementsCount()
         /// The id of the achievement to unlock for the current player.
         /// </param>
-        public Task UnlockAchievementAsync(string achievementId)
+        public Task<PlayerAchievement> UnlockAchievementAsync(string achivementId)
```
The name of this param has a spelling mistake

JessTello replied:
Fixed in the latest commit.

JessTello committed:
Update AchievementsService.cs

- Deleted unused struct
- Corrected spelling error in variable name

matt-clarke committed:
Minor adjustments to AchievementsService

- Use nullable for `definition` var
- Cache achievement lookup in first loop to avoid double lookup
- Fix indentation/line breaks in lambda

billfr-ApexSystems reviewed (approved):
Approved with changes requested by Matt
